### ISSUE TITLE
Consistency in array-length limits

### ISSUE THREAD

**Issue by RustanLeino** [dafny-lang/dafny#3212]
### Summary

Most of our compiler targets check some limit on the size of arrays that can be allocated. If the limit is exceeded, the program halts with some message like "out of memory" or (perhaps more preferably) "array-length limit exceeded". In return, this lets the runtimes use the corresponding word size when dealing with array lengths and array indices.

We have spent some effort to try to make this consistent across the compilers, but more can be done. In some places, this check is done twice, and in some places, the error string is repeated once for each check. Also, the code generally does not rely on the fact that indices are word sizes.

This Issue is requesting an overhaul of this functionality, to make sure that it is consistent and documented.

### Background and Motivation

See above.

### Proposed Feature

See above.

**Comment by robin-aws** [comment]
Assigning myself as https://github.com/dafny-lang/dafny/pull/3167 already makes some progress in this direction: implementing the sequence type soundly in Dafny source code forces the issue.

@RustanLeino just to clarify, I assume you're suggesting that we allow each backend to independently decide what their size limit is, rather than picking a single limit across backends that we would document as Dafny's size limit?

**Comment by davidcok** [comment]
And I would concur with the sentiment you are attributing to Rustan. In fact I think it is fair to document across the board the different limitations of each backend platform, even if they all are less than what the Dafny provers can handle. Ideally we can warn about some of them. But resource limitations will be affected also by the user's own system.

**Comment by robin-aws** [comment]
Agreed. And I wouldn't even call this a "limitation", it's an intentional tradeoff. 

We could allow arrays to be arbitrarily large, constrained only by the actual memory limits, if we always used native `BigInteger`s as array indexes and used chains of native arrays to implement arrays longer than the maximum native type. But that would slow down the common case quite a bit so it's probably not the right choice. At best we could implement an adaptive approach where we dynamically switch between the implementations somehow, much as Python supports arbitrarily large integers but optimizes for smaller values, but that's quite a bit of complexity to add.
